### DISCUSSION TITLE
Get Icon of a file

### DISCUSSION THREAD

**Discussion by tonka3000** [halildurmus/win32#676] [General]
Hi,

does anybody have a snippet how to get an icon of any file (like `.exe`, `.lnk`). AFAIK [SHGetImageList](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shgetimagelist) is the key, but I'm not able to call this correctly (new to win32 dart package).

Thanks in advance

**Comment by fitewzo** [comment]
import 'dart:ffi';
import 'dart:io';

import 'package:ffi/ffi.dart';
import 'package:win32/win32.dart';

void main() {
  final filePath = 'C:\\Windows\\notepad.exe';

  final hIcon = allocate<IntPtr>();

  final result = SHGetFileInfo(
    filePath.toNativeUtf16(),
    FILE_ATTRIBUTE_NORMAL,
    allocate<SHFILEINFO>().addressOf,
    sizeOf<SHFILEINFO>(),
    SHGFI_ICON | SHGFI_SMALLICON | SHGFI_USEFILEATTRIBUTES,
  );

  if (result != 0) {
    final iconHandle = SHGetStockIconInfo(
      SHSTOCKICONID.SIID_APPLICATION,
      SHGSI_ICONLOCATION | SHGSI_SMALLICON | SHGSI_SYSICONINDEX,
    );
    final hInstance = GetModuleHandle(nullptr);
    final iconIndex = LookupIconIdFromDirectoryEx(
      iconHandle.hIcon,
      true,
      GetSystemMetrics(SystemMetric.SM_CXICON),
      GetSystemMetrics(SystemMetric.SM_CYICON),
      LR_DEFAULTCOLOR,
    );

    ExtractIconEx(
      iconHandle.szPath,
      iconIndex,
      nullptr,
      hIcon,
      1,
    );

    DestroyIcon(iconHandle.hIcon);

    final iconPath = '${Directory.current.path}\\icon.ico';
    final hFile = CreateFile(
      iconPath.toNativeUtf16(),
      GENERIC_WRITE,
      0,
      nullptr,
      CREATE_ALWAYS,
      FILE_ATTRIBUTE_NORMAL,
      nullptr,
    );

    final pIconData = allocate<Uint8>(count: 4096);
    final iconSize = GetIconInfo(hIcon.value, pIconData);

    if (iconSize > 0) {
      final written = allocate<Uint32>();

      WriteFile(hFile, pIconData, iconSize, written, nullptr);

      Free(pIconData);
      CloseHandle(hFile);

      print('Icon saved to $iconPath');
    } else {
      print('Error getting icon information');
    }
  } else {
    print('Error retrieving file information');
  }
}

**Comment by tonka3000** [comment]
Thank you very much @fitewzo , I will try it out.
